core/src/main/java/org/acegisecurity/BadCredentialsException.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2020 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.acegisecurity;
+
+import org.acegisecurity.userdetails.UsernameNotFoundException;
+
+/**
+ * @deprecated use {@link org.springframework.security.authentication.BadCredentialsException}
+ */
+@Deprecated
```

Comment:
would it be possible to `@Restrict(NoExternalUse.class)` this (so that forces plugins to upgrade).
perhaps that is a general question that should be asked higher up?

Reply:
That is a general question and my answer is emphatically no. `@Deprecated` means it will continue to work but you should switch to the new thing as soon as you can. That is exactly the case. Abuse of `@Restricted` has caused terrible headaches.